Apple Clang prioritises /usr/local/include over CMake SYSTEM includes

## Problem Statement

Apple clang includes `/usr/local/include` as a non-system include path.

```console
$ c++ -E -v - </dev/null
Apple clang version 16.0.0 (clang-1600.0.26.6)
Target: arm64-apple-darwin24.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple arm64-apple-macosx15.0.0 -Wundef-prefix=TARGET_OS_ -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -E -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -mframe-pointer=non-leaf -fno-strict-return -ffp-contract=on -fno-rounding-math -funwind-tables=1 -fobjc-msgsend-selector-stubs -target-sdk-version=15.5 -fvisibility-inlines-hidden-static-local-var -fno-modulemap-allow-subdirectory-search -target-cpu apple-m1 -target-feature +v8.5a -target-feature +aes -target-feature +crc -target-feature +dotprod -target-feature +fp-armv8 -target-feature +fp16fml -target-feature +lse -target-feature +ras -target-feature +rcpc -target-feature +rdm -target-feature +sha2 -target-feature +sha3 -target-feature +neon -target-feature +zcm -target-feature +zcz -target-feature +fullfp16 -target-abi darwinpcs -debugger-tuning=lldb -target-linker-version 1115.7.3 -v -fcoverage-compilation-dir=/Users/fahien/Workspace/lunar/gfxreconstruct -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -Wno-elaborated-enum-base -Wno-reserved-identifier -Wno-gnu-folding-constant -fdebug-compilation-dir=/Users/fahien/Workspace/lunar/gfxreconstruct -ferror-limit 19 -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fmax-type-align=16 -fcommon -fcolor-diagnostics -clang-vendor-feature=+disableNonDependentMemberExprInCurrentInstantiation -fno-odr-hash-protocols -clang-vendor-feature=+enableAggressiveVLAFolding -clang-vendor-feature=+revert09abecef7bbf -clang-vendor-feature=+thisNoAlignAttr -clang-vendor-feature=+thisNoNullAttr -clang-vendor-feature=+disableAtImportPrivateFrameworkInImplementationError -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o - -x c -
clang -cc1 version 16.0.0 (clang-1600.0.26.6) default target arm64-apple-darwin24.5.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
# 1 "<stdin>"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 424 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "<stdin>" 2
```

Our `CMakeLists.txt` specify this for `vulkan_registry`:
```cmake
target_include_directories(vulkan_registry SYSTEM INTERFACE ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
```

I suppose `SYSTEM` has been included to avoid compilation warnings for https://github.com/LunarG/gfxreconstruct/pull/1891

Problem is that the compiler has the following order for include directories:

```console
$ c++ -isystem /Users/fahien/Workspace/lunar/gfxreconstruct/external/Vulkan-Headers/include -Wp,-v -E -
clang -cc1 version 16.0.0 (clang-1600.0.26.6) default target arm64-apple-darwin24.5.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Users/fahien/Workspace/lunar/gfxreconstruct/external/Vulkan-Headers/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

When the VulkanSDK is installed on MacOS, headers can be found in `/usr/local/include`. This means the compiler will find `vulkan/vulkan_core.h` in `/usr/local/include` instead of `/Users/fahien/Workspace/lunar/gfxreconstruct/external/Vulkan-Headers/include/external/Vulkan-Headers/include` as we expect.

The VulkanSDK version installed might be older than the `gfxreconstruct/external/Vulkan-Headers` version directory. The code generator is using `gfxreconstruct/external/Vulkan-Headers` to generate code, but the compiler is including the headers from `/usr/local/include` which results in a compiler error for new symbols generated which are not available in older version of the headers. E.g.:

```c++
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/util/thread_data.cpp:23:
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/util/thread_data.h:28:
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/encode/parameter_encoder.h:29:
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/encode/openxr_handle_wrapper_util.h:33:
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/generated/generated_openxr_state_table.h:37:
In file included from /Users/fahien/Workspace/lunar/gfxreconstruct/framework/encode/vulkan_handle_wrappers.h:31:
/Users/fahien/Workspace/lunar/gfxreconstruct/framework/generated/generated_vulkan_dispatch_table.h:486:92: error: unknown type name 'VkPresentWait2InfoKHR'; did you mean 'VkPresentInfoKHR'?
  486 | inline VKAPI_ATTR VkResult VKAPI_CALL vkWaitForPresent2KHR(VkDevice, VkSwapchainKHR, const VkPresentWait2InfoKHR*) { GFXRECON_LOG_WARNING_ONCE("Unsupported function vkWaitForPresent2KHR was called, resulting in no-op behavior."); return VK_SUCCESS; }
      |                                                                                            ^~~~~~~~~~~~~~~~~~~~~
      |                                                                                            VkPresentInfoKHR
/usr/local/include/vulkan/vulkan_core.h:8713:3: note: 'VkPresentInfoKHR' declared here
 8713 | } VkPresentInfoKHR;
 ```

## Current Workaround

Use llvm's clang.

```console
$ brew install llvm
```

Configure the project with these settings:

```console
$ cmake \
     -DCMAKE_C_COMPILER=`brew --prefix llvm`/bin/clang \
     -DCMAKE_CXX_COMPILER=`brew --prefix llvm`/bin/clang++ \
     ...
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -414,7 +414,15 @@ if(MSVC)
 endif()
 
 add_library(vulkan_registry INTERFACE)
-target_include_directories(vulkan_registry SYSTEM INTERFACE ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
+if (CMAKE_CXX_COMPILER_ID STREQUAL AppleClang)
+    # AppleClang implicitly adds /usr/local/include as a -I path, which takes precedence
+    # over -isystem paths. Add external/Vulkan-Headers as a normal -I path for the correct
+    # search order, then use --system-header-prefix to treat it as a system header for warnings
+    target_include_directories(vulkan_registry INTERFACE ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
+    target_compile_options(vulkan_registry INTERFACE --system-header-prefix vulkan/)
+else ()
+    target_include_directories(vulkan_registry SYSTEM INTERFACE ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
+endif()
 target_compile_definitions(vulkan_registry INTERFACE VK_NO_PROTOTYPES VK_ENABLE_BETA_EXTENSIONS)
 
 add_library(spirv_registry INTERFACE)

